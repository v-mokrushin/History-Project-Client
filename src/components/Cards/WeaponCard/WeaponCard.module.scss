@import "/src/assets/css/common.scss";

.root {
  aspect-ratio: 1.65 / 1;
  width: 100%;
  border-radius: 1.5%;
  position: relative;
  text-decoration: none;
  transition: transform 300ms ease;
  display: flex;
  justify-content: center;
  align-items: flex-start;
  box-shadow: 0px 0px 10px 0px rgba(36, 39, 42, 0.3);
  overflow: hidden;
  background-color: #000000;

  @include breakpoint(md) {
    border-radius: 2%;
    box-shadow: 0px 0px 4px 0px rgba(36, 39, 42, 0.3);
  }

  &_large %text {
    @include breakpoint(xs) {
      font-size: 14px;
    }
  }

  &::after {
    position: absolute;
    content: "";
    background: linear-gradient(
      to bottom,
      rgba(0, 0, 0, 0.15) 0%,
      rgba(0, 0, 0, 0.2) 65%,
      #000000 100%
    );
    left: 0;
    right: 0;
    top: 0;
    bottom: 0;
    transition: background-color 300ms ease;
  }

  &:hover::after {
    background-color: rgba($color: #000000, $alpha: 0.25);
  }

  &:hover .description {
    opacity: 1;
  }

  &:hover .title {
    color: $gold;
    transform: translateY(0);
  }

  &:hover .background {
    transform: scale(1.06);
  }
}

.background {
  position: absolute;
  // left: 0;
  // top: 0;
  // right: 0;
  // bottom: 0;
  height: 100%;
  width: 100%;
  background-repeat: no-repeat;
  background-position: center;
  background-size: cover;
  transition: transform 300ms ease;
  z-index: 0;
  will-change: transform;
}

.video {
  position: absolute;
  height: 100%;
  width: 100%;
  object-fit: cover;
  z-index: 0;
  will-change: transform;
}

.contentWrapper {
  height: 100%;
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: flex-end;
  gap: 2px;
  position: relative;
  padding: 10px;
  z-index: 3;

  @include breakpoint(FHD) {
    padding: 9px;
  }

  @include breakpoint(xl) {
    padding: 7px;
  }

  @include breakpoint(sm) {
    gap: 10px;
    padding: 8px;
  }

  @include breakpoint(xs) {
    padding: 4px;
    gap: 5px;
  }
}

%text {
  font-family: "Rubik", sans-serif;
  font-weight: 500;
  font-size: 16px;
  color: $white;
  letter-spacing: 1px;
  transition: all 300ms ease;
  text-align: center;
  font-display: fallback;

  @include breakpoint(FHD) {
    font-size: 15px;
  }

  @include breakpoint(xl) {
    font-size: 12px;
  }

  @include breakpoint(xs) {
    font-size: 10px;
  }
}

.title {
  @extend %text;
  transform: translateY(22px);

  @include breakpoint(xl) {
    transform: translateY(15px);
  }

  @include breakpoint(lg) {
    transform: translateY(17px);
  }

  @include breakpoint(md) {
    transform: translateY(0);
  }
}

.description {
  @extend %text;
  opacity: 0;
  transition: opacity 200ms ease;

  @include breakpoint(md) {
    display: none;
  }
}
