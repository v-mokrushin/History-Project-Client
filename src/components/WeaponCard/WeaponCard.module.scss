@import "/src/assets/css/common.scss";

.root {
  aspect-ratio: 1.65 / 1;
  width: 100%;
  border-radius: 2px;
  position: relative;
  text-decoration: none;
  transition: transform 300ms ease;
  display: flex;
  justify-content: center;
  align-items: flex-end;
  box-shadow: 0px 0px 15px 0px rgba(36, 39, 42, 0.3);
  overflow: hidden;

  @include media-breakpoint-down(md) {
    box-shadow: 0px 0px 4px 0px rgba(36, 39, 42, 0.3);
  }

  &_large %text {
    @include media-breakpoint-down(xs) {
      font-size: 14px;
    }
  }

  &::after {
    position: absolute;
    content: "";
    background: linear-gradient(
      to bottom,
      rgba(0, 0, 0, 0.15) 0%,
      rgba(0, 0, 0, 0.2) 65%,
      #000000 100%
    );
    left: 0;
    right: 0;
    top: 0;
    bottom: 0;
    transition: background-color 300ms ease;
  }

  &:hover::after {
    background-color: rgba($color: #000000, $alpha: 0.25);
  }

  &:hover .description {
    opacity: 1;
  }

  &:hover .title {
    color: $primary-gold-text-color;
    transform: translateY(0);
  }

  &:hover .backgroundWrapper {
    transform: scale(1.06);
  }
}

.backgroundWrapper {
  position: absolute;
  // left: 0;
  // top: 0;
  // right: 0;
  // bottom: 0;
  height: 100%;
  width: 100%;
  background-repeat: no-repeat;
  background-position: center center;
  background-size: cover;
  transition: transform 300ms ease;
  z-index: 0;
}

.contentWrapper {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 2px;
  position: relative;
  padding: 10px;
  z-index: 3;

  @include media-breakpoint-down(FHD) {
    padding: 9px;
  }

  @include media-breakpoint-down(xl) {
    padding: 7px;
  }

  @include media-breakpoint-down(sm) {
    gap: 10px;
    padding: 8px;
  }

  @include media-breakpoint-down(xs) {
    padding: 4px;
    gap: 5px;
  }
}

%text {
  font-family: "Rubik", sans-serif;
  font-weight: 500;
  font-size: 19px;
  color: $primary-light-text-color;
  letter-spacing: 1px;
  transition: all 300ms ease;
  text-align: center;

  @include media-breakpoint-down(FHD) {
    font-size: 16px;
  }

  @include media-breakpoint-down(xl) {
    font-size: 13px;
  }

  @include media-breakpoint-down(lg) {
  }

  @include media-breakpoint-down(md) {
  }

  @include media-breakpoint-down(sm) {
    font-size: 12px;
  }

  @include media-breakpoint-down(xs) {
    font-size: 10px;
  }
}

.title {
  @extend %text;
  transform: translateY(22px);

  @include media-breakpoint-down(xl) {
    transform: translateY(15px);
  }

  @include media-breakpoint-down(lg) {
    transform: translateY(17px);
  }

  @include media-breakpoint-down(md) {
    transform: translateY(0);
  }
}

.description {
  @extend %text;
  opacity: 0;
  transition: opacity 200ms ease;

  @include media-breakpoint-down(md) {
    display: none;
  }
}
