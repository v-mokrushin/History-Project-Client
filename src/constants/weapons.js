let armoredVehicle = {
  name: {
    russian: "Бронетехника",
    english: "armored vehicles",
    path: "armored-vehicles",
  },
  lightTank: {
    name: {
      russian: "Лёгкий танк",
    },
  },
  mediumTank: {
    name: {
      russian: "Средний танк",
    },
  },
  heavyTank: {
    name: {
      russian: "Тяжелый танк",
    },
  },
  sau: {
    name: {
      russian: "САУ",
    },
  },
  ptsau: {
    name: {
      russian: "ПТ-САУ",
    },
  },
  armoredCar: {
    name: {
      russian: "Бронеавтомобиль",
    },
  },
};

let aviation = {
  name: {
    russian: "Авиация",
    english: "aviation",
    path: "aviation",
  },
  fighter: {
    name: {
      russian: "Истребитель",
    },
  },
  interceptor: {
    name: {
      russian: "Перехватчик",
    },
  },
  attackАircraft: {
    name: {
      russian: "Штурмовик",
    },
  },
  bomber: {
    name: {
      russian: "Бомбардировщик",
    },
  },
};

let artillery = {
  name: {
    russian: "Артиллерия",
    english: "artillery",
    path: "artillery",
  },
  antitank: {
    name: {
      russian: "Противотанковая",
    },
  },
  division: {
    name: {
      russian: "Дивизионная",
    },
  },
  armyAndCorps: {
    name: {
      russian: "Корпусная/армейская",
    },
  },
  specialPower: {
    name: {
      russian: "Большой и особой мощности",
    },
  },
  antiaircraft: {
    name: {
      russian: "Зенитная",
    },
  },
};

let smallArms = {
  name: {
    russian: "Стрелковое оружие",
    english: "small arms",
    path: "small-arms",
  },
  revolver: {
    name: {
      russian: "Револьвер",
    },
  },
  gun: {
    name: {
      russian: "Пистолет",
    },
  },
  rifle: {
    name: {
      russian: "Винтовка и карабин",
    },
  },
  selfLoadingRifle: {
    name: {
      russian: "Самозарядная  винтовка",
    },
  },
  assaultRifle: {
    name: {
      russian: "Штурмовая винтовка",
    },
  },
  machineGun: {
    name: {
      russian: "Пулемет",
    },
  },
  antiTankGun: {
    name: {
      russian: "Противотанковое ружье",
    },
  },
};

appendBaseType(armoredVehicle);
appendBaseType(aviation);
appendBaseType(smallArms);
appendBaseType(artillery);

export const WEAPONS_TYPE = {
  armoredVehicle,
  aviation,
  smallArms,
  artillery,
  getObjectByPath: getByPath,
  getTypesArray,
  getTypesArrayWithAll,
  getAllType,
  isAllType,
};

const allType = {
  name: {
    russian: "Все",
  },
};

function getByPath(path) {
  for (let key in WEAPONS_TYPE) {
    if (WEAPONS_TYPE[key].name.path === path) {
      return WEAPONS_TYPE[key];
    }
  }
}

function appendBaseType(object) {
  for (let [key, value] of Object.entries(object)) {
    if (key != "name") value.baseType = object.name;
  }
}

function getAllType() {
  return allType;
}

function getTypesArray(weaponBranch) {
  const copy = {};
  Object.assign(copy, weaponBranch);
  delete copy.name;
  return Object.values(copy);
}

function isAllType(type) {
  return getAllType() === type;
}

function getTypesArrayWithAll(weaponBranch) {
  return [getAllType(), ...getTypesArray(weaponBranch)];
}
