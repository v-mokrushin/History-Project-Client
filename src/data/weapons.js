import { ARMORED_VEHICLES } from "./armored";
import { AVIATION_DATA } from "./aviation";

export const WEAPONS_DATA = [...ARMORED_VEHICLES, ...AVIATION_DATA];
