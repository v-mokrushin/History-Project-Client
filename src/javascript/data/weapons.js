import { ARMORED_VEHICLES } from "./armored";
import { ARTILLERY_DATA } from "./artillery";
import { AVIATION_DATA } from "./aviation";
import { SMALL_ARMS_DATA } from "./smallArms";

export const WEAPONS_DATA = [
  ...ARMORED_VEHICLES,
  ...AVIATION_DATA,
  ...ARTILLERY_DATA,
  ...SMALL_ARMS_DATA,
];

WEAPONS_DATA.forEach((item) => {
  let name = item.name;
  if (name.at(-1) === ".") name = name.slice(0, -1);

  Object.defineProperty(item, "id", {
    get: function () {
      return this.name.replaceAll(" ", "-").replaceAll("/", "-");
    },
  });

  item.gallery = {
    icon:
      `/images/weapons/${item.type.branch.path}/` +
      item.nation.path +
      "/" +
      name
        .replaceAll(" ", "-")
        .replaceAll("/", "-")
        .replaceAll("«", "")
        .replaceAll("»", "")
        .replaceAll("(", "")
        .replaceAll(")", "") +
      "/" +
      item.icon,
  };

  delete item.icon;
});
