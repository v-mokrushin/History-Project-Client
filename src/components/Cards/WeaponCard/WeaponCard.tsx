import React from "react";
import { NavLink } from "react-router-dom";
import styles from "./WeaponCard.module.scss";
import classNames from "classnames";
import { ANIMATIONS } from "../../../constants/animations";
import { getBackgroundImageStyleObject } from "../../../utils/common";
import Flag from "../../Graphics/Flag/Flag";
import { observer } from "mobx-react";
import settingsStore from "stores/mobx/settingsStore";
import { TWeapon } from "data/weapons/interfaces/common-weapon-interfaces";
import Preloader from "components/Graphics/Preloader/Preloader";

interface IWeaponCardProps {
  weapon: TWeapon;
  showFlag?: boolean;
  isAbsoluteLinkPath?: boolean;
  className?: string;
}

const WeaponCard = observer(
  ({
    weapon,
    showFlag = false,
    isAbsoluteLinkPath = false,
    className,
  }: IWeaponCardProps) => {
    const colorized = settingsStore.colorized;

    return (
      <NavLink
        to={
          isAbsoluteLinkPath
            ? `/weapons/${weapon.branch?.path}/world/${weapon.id}`
            : weapon.id!
        }
        className={classNames(
          styles.root,
          settingsStore.displaySize && styles.root_large,
          ANIMATIONS.fadeIn,
          className
        )}
        key={Math.random()}
      >
        <div
          className={styles.background}
          style={getBackgroundImageStyleObject(weapon.gallery?.icon)}
        ></div>
        <Preloader color="white" />
        {showFlag && weapon.nation && (
          <Flag nation={weapon.nation} minimized isNavLink={false} />
        )}
        <div className={styles.contentWrapper}>
          <p className={styles.title}>{weapon?.name}</p>
          <p className={styles.description}>{weapon?.type.name.russian}</p>
        </div>
      </NavLink>
    );
  }
);

export default WeaponCard;
