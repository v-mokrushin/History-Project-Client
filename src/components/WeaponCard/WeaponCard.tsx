import React from "react";
import { NavLink } from "react-router-dom";
import styles from "./WeaponCard.module.scss";
import classNames from "classnames";
import { ANIMATIONS } from "../../constants/animations";
import { getBackgroundImageStyleObject } from "../../utils/common";
import Flag from "../Flag/Flag";
import { observer } from "mobx-react";
import settingsStore from "stores/mobx/settingsStore";
import { TWeapon } from "data/weapons/interfaces/common-weapon-interfaces";
import { Scroll } from "utils/scroll";

interface IWeaponCardProps {
  weapon: TWeapon;
  showFlag?: boolean;
  isAbsoluteLinkPath?: boolean;
}

const WeaponCard = observer(
  ({
    weapon,
    showFlag = false,
    isAbsoluteLinkPath = false,
  }: IWeaponCardProps) => {
    const colorized = settingsStore.colorized;

    return (
      <NavLink
        to={
          isAbsoluteLinkPath
            ? `/weapons/${weapon.branch?.path}/world/${weapon.id}`
            : weapon.id!
        }
        className={classNames(
          styles.root,
          settingsStore.displaySize && styles.root_large,
          ANIMATIONS.fadeIn
        )}
        onClick={() => {
          Scroll.toTopInstantly();
        }}
      >
        <div
          className={styles.backgroundWrapper}
          style={getBackgroundImageStyleObject(weapon.gallery?.icon)}
        ></div>
        {showFlag && weapon.nation && (
          <Flag nation={weapon.nation} minimized isNavLink={false} />
        )}
        <div className={styles.contentWrapper}>
          <p className={styles.title}>{weapon?.name}</p>
          <p className={styles.description}>{weapon?.type.name.russian}</p>
        </div>
      </NavLink>
    );
  }
);

export default WeaponCard;
