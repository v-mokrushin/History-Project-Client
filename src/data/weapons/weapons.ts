import { IChiefDesigner } from "./departments/chief-designers";
import { IDeveloper } from "./departments/developers";
import { shuffleArray } from "utils/common";
import {
  defineGallery,
  defineIdProperty,
  defineModels,
  definePathProperty,
  prepareWeapon,
  prepareWeapons,
  readWeaponsFromLocalStorage,
  sortByTitle,
} from "utils/weapons";
import { INation, Nations } from "../../constants/nations";
import { ARMORED_VEHICLES } from "./branches/armored-vehicles";
import { ARTILLERY_DATA } from "./branches/artillery";
import { AVIATION_DATA } from "./branches/aviation";
import { GRENADE_LAUNCHERS_DATA } from "./branches/grenade-launchers";
import { SMALL_ARMS_DATA } from "./branches/small-arms";
import { TFilters } from "stores/mobx/filtersStore";
import { TWeapon } from "./interfaces/common-weapon-interfaces";
import { ISelectionVariantWithFlag } from "components/Controls/Filter/Filter";
import { IProducer } from "./departments/producers";
import { Random } from "utils/random";
import { NationsMethods } from "../../constants/nations";

const weapons_data = ([] as TWeapon[]).concat(
  ARMORED_VEHICLES,
  AVIATION_DATA,
  ARTILLERY_DATA,
  SMALL_ARMS_DATA,
  GRENADE_LAUNCHERS_DATA
);

prepareWeapons(weapons_data);
readWeaponsFromLocalStorage(weapons_data);

// --------------------------------------------------------------------------------

export class Weapons {
  public static addNewWeapon(weapon: TWeapon) {
    weapons_data.unshift(weapon);
  }

  public get() {
    return weapons_data;
  }

  public static doesWeaponExist(name: string): boolean {
    return weapons_data.find((weapon) => weapon.name === name) ? true : false;
  }

  public static getWeaponExample(): TWeapon {
    return weapons_data[0];
  }

  public static getUniqueDates(collection: TWeapon[]): number[] {
    let dates = collection.map((item) => item.adoptedIntoServiceDate);
    dates = Array.from(new Set(dates)).sort((a, b) => b - a);
    return dates;
  }

  public static selectWeapons(
    weaponBranchPath: string | undefined,
    nationPath: string | undefined
  ): TWeapon[] {
    return weapons_data
      .filter((item) => item.branch?.path === weaponBranchPath)
      .filter(
        (item) =>
          nationPath === Nations.World.path || item.nation!.path === nationPath
      );
  }

  public static filterByName(name: string) {
    return weapons_data.filter((item) =>
      item.name.toLowerCase().includes(String(name).toLowerCase())
    );
  }

  public static filterWeapons(
    selectedWeapons: TWeapon[],
    filters: TFilters
  ): TWeapon[] {
    if (Object.keys(filters).length === 0) {
      return selectedWeapons;
    } else {
      let weapons = [...selectedWeapons];

      if (filters.name)
        weapons = weapons.filter((item) =>
          item.name.toLowerCase().includes(String(filters.name).toLowerCase())
        );

      if (filters.nation)
        weapons = weapons.filter(
          (item) => item.nation?.name.russian === filters.nation
        );

      if (filters.type)
        weapons = weapons.filter(
          (item) => item.type.name.russian === filters.type
        );

      if (filters.developer)
        weapons = weapons.filter(
          (weapon) =>
            weapon.specifications?.common?.developer?.name.original ==
            filters.developer
        );

      if (filters.chiefDesigner)
        weapons = weapons.filter(
          (weapon) =>
            weapon.specifications?.common?.chiefDesigner?.name.russian ==
            filters.chiefDesigner
        );

      if (filters.platform)
        weapons = weapons.filter(
          (weapon) =>
            weapon.specifications?.common?.platform?.name.original ==
            filters.platform
        );

      if (filters.producer)
        weapons = weapons.filter((weapon) =>
          weapon.specifications?.common?.producer?.find(
            (item) => item.name.original === filters.producer
          )
        );

      return weapons;
    }
  }

  public static selectNation(weaponsBranchPath: string | undefined) {
    return [
      Nations.World,
      ...Array.from(
        new Set(
          weapons_data
            .filter((item) => item.branch?.path === weaponsBranchPath)
            .map((item) => item.nation)
        )
      ),
    ];
  }

  public static getById(weaponId: string | undefined): TWeapon | undefined {
    return weapons_data.find((item) => item.id === weaponId);
  }

  public static changeColorized(): void {
    weapons_data.forEach((weapon) => {
      if (weapon.gallery) {
        weapon.gallery.isColorizedIcon = !weapon.gallery.isColorizedIcon;
      }
    });
  }

  public static getRandomWeapon(): TWeapon {
    return weapons_data[Random.getInteger(0, weapons_data.length) - 1];
  }

  public static getRecommendation(
    weaponBranchPath: string | undefined,
    weaponId: string | undefined
  ): TWeapon[] {
    let weapons = weapons_data.filter(
      (weapon) =>
        weapon.branch?.path === weaponBranchPath && weapon.id !== weaponId
    );
    weapons = shuffleArray(weapons);
    weapons = weapons.slice(0, 8);

    return weapons;
  }

  // getTypes(selectedWeapons: TWeapon[]) {
  //   return Object.values(weaponBranch)
  //   .map((item) => item.name?.russian)
  //   .filter((item) => item != undefined);
  // },

  public static getDevelopers(selectedWeapons: TWeapon[]): string[] {
    let developers: string[] = selectedWeapons
      .filter((weapon) => {
        if (!weapon.specifications.common.developer) return false;
        return weapon.specifications.common.developer;
      })
      .map((weapon) => weapon.specifications.common.developer!.name.original);

    developers = Array.from(new Set(developers));
    developers.sort().unshift("Все");
    return developers;
  }

  public static getDevelopersWithFlags(
    selectedWeapons: TWeapon[]
  ): ISelectionVariantWithFlag[] {
    let developers: IDeveloper[] = selectedWeapons
      .filter((weapon) => {
        return weapon.specifications.common.developer;
      })
      .map((item) => item.specifications.common.developer!);
    developers = Array.from(new Set(developers));

    let variants: ISelectionVariantWithFlag[] = developers.map((item) => ({
      title: item.name.original,
      nation: item.nation,
    }));

    sortByTitle(variants);
    variants.unshift({ title: "Все", nation: Nations.World });

    return variants;
  }

  public static getNations(selectedWeapons: TWeapon[]): string[] {
    let nations: string[] = selectedWeapons.map(
      (weapon) => weapon.nation?.name.russian!
    );

    nations = Array.from(new Set(nations));
    nations.sort().unshift("Весь мир");
    return nations;
  }

  public static getNationsWithFlags(
    selectedWeapons: TWeapon[]
  ): ISelectionVariantWithFlag[] {
    let nations: INation[] = selectedWeapons
      .filter((weapon) => {
        return weapon.nation;
      })
      .map((item) => item.nation!);
    nations = Array.from(new Set(nations));

    let variants: ISelectionVariantWithFlag[] = nations.map((item) => ({
      title: item.name.russian,
      nation: item,
    }));

    sortByTitle(variants);
    variants.unshift({ title: "Весь мир", nation: Nations.World });

    return variants;
  }

  public static getChiefDesigners(selectedWeapons: TWeapon[]): string[] {
    let designers: string[] = selectedWeapons
      .filter((weapon) => {
        return weapon.specifications.common.chiefDesigner;
      })
      .map(
        (weapon) => weapon.specifications.common.chiefDesigner!.name.russian
      );

    designers = Array.from(new Set(designers));
    designers.sort().unshift("Все");
    return designers;
  }

  public static getChiefDesignersWithFlags(
    selectedWeapons: TWeapon[]
  ): ISelectionVariantWithFlag[] {
    let designers: IChiefDesigner[] = selectedWeapons
      .filter((weapon) => {
        return weapon.specifications.common.chiefDesigner;
      })
      .map((item) => item.specifications.common.chiefDesigner!);
    designers = Array.from(new Set(designers));

    let variants: ISelectionVariantWithFlag[] = designers.map((item) => ({
      title: item.name.russian,
      nation: item.nation,
    }));

    sortByTitle(variants);
    variants.unshift({ title: "Все", nation: Nations.World });

    return variants;
  }

  public static getPlatforms(selectedWeapons: TWeapon[]): string[] {
    let platforms: string[] = selectedWeapons
      .filter((weapon) => {
        if (!weapon.specifications) return false;
        if (!weapon.specifications.common) return false;
        return weapon.specifications.common.platform;
      })
      .map((weapon) => weapon.specifications.common.platform!.name.original);

    platforms = Array.from(new Set(platforms));
    platforms.sort().unshift("Все");
    return platforms;
  }

  public static getProducersWithFlags(
    selectedWeapons: TWeapon[]
  ): ISelectionVariantWithFlag[] {
    let producers: IProducer[] = selectedWeapons
      .filter((weapon) => {
        return weapon.specifications.common.producer;
      })
      .map((item) => item.specifications.common.producer!)
      .flat();

    producers = Array.from(new Set(producers));

    let variants: ISelectionVariantWithFlag[] = producers.map((item) => ({
      title: item.name.original,
      nation: item.nation,
    }));

    sortByTitle(variants);
    variants.unshift({ title: "Все", nation: Nations.World });

    return variants;
  }
}

// console.log(weapons_data);
