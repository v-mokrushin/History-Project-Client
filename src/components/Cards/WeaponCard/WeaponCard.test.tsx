import { render } from "@testing-library/react";
import { Weapons } from "data/weapons/weapons";
import { BrowserRouter } from "react-router-dom";
import WeaponCard from "./WeaponCard";

describe("WeaponCard", () => {
  test("WeaponCard render", () => {
    const weaponCard = render(
      <BrowserRouter>
        <WeaponCard weapon={Weapons.getWeaponExample()} showFlag />
      </BrowserRouter>
    );
    expect(weaponCard).toMatchSnapshot();
  });
});
