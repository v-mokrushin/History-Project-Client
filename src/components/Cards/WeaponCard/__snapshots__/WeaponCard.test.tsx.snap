// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeaponCard WeaponCard render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <a
        class="root animate__animated animate__fadeIn"
        href="/ИС-3"
      >
        <div
          class="background"
          style="background-image: url(/images/weapons/armored-vehicles/soviet-union/ИС-3/icon.jpg);"
        />
        <div
          class="root white"
        />
        <div
          class="root minimized"
        >
          <div
            class="flag"
            style="background-image: url(/images/flags/flag-ussr.webp);"
          />
        </div>
        <div
          class="contentWrapper"
        >
          <p
            class="title"
          >
            ИС-3
          </p>
          <p
            class="description"
          >
            Тяжелый танк
          </p>
        </div>
      </a>
    </div>
  </body>,
  "container": <div>
    <a
      class="root animate__animated animate__fadeIn"
      href="/ИС-3"
    >
      <div
        class="background"
        style="background-image: url(/images/weapons/armored-vehicles/soviet-union/ИС-3/icon.jpg);"
      />
      <div
        class="root white"
      />
      <div
        class="root minimized"
      >
        <div
          class="flag"
          style="background-image: url(/images/flags/flag-ussr.webp);"
        />
      </div>
      <div
        class="contentWrapper"
      >
        <p
          class="title"
        >
          ИС-3
        </p>
        <p
          class="description"
        >
          Тяжелый танк
        </p>
      </div>
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
